packages/core/test/microstep.test.ts

```diff
@@ -44,15 +81,105 @@ describe('machine.microstep()', () => {
       }
     });
 
-    const state = machine.microstep('first', 'TRIGGER');
-
-    expect(state.matches('second')).toBeTruthy();
-    expect(state._internalQueue).toContainEqual(
-      expect.objectContaining(
-        toSCXMLEvent({
-          type: 'RAISED'
-        })
-      )
+    const states = machine.microstep(
+      machine.resolveStateValue('first'),
+      'TRIGGER',
+      undefined
     );
+
+    expect(states.map((s) => s.value)).toEqual(['second', 'third']);
+  });
+
+  it('should return a single-item array for normal transitions', () => {
+    const machine = createMachine({
+      initial: 'first',
+      states: {
+        first: {
+          on: {
+            TRIGGER: 'second'
+          }
+        },
+        second: {}
+      }
+    });
+
+    const states = machine.microstep(machine.initialState, 'TRIGGER');
+
+    expect(states.map((s) => s.value)).toEqual(['second']);
+  });
+
+  it('each state should preserve their internal queue', () => {
+    const machine = createMachine({
+      initial: 'first',
+      states: {
+        first: {
+          on: {
+            TRIGGER: {
+              target: 'second',
+              actions: [raise('FOO'), raise('BAR')]
+            }
+          }
+        },
+        second: {
+          on: {
+            FOO: {
+              target: 'third'
+            }
+          }
+        },
+        third: {
+          on: {
+            BAR: {
+              target: 'fourth'
+            }
+          }
+        },
+        fourth: {
+          always: 'fifth'
+        },
+        fifth: {}
+      }
+    });
+
+    const states = machine.microstep(machine.initialState, 'TRIGGER');
+
+    expect(states.map((s) => [s.value, s._internalQueue.length])).toEqual([
+      ['second', 2], // foo, bar
+      ['third', 1], // bar
+      ['fourth', 0], // (eventless)
+      ['fifth', 0]
+    ]);
+  });
+
+  it('actions are preserved throughout microstep', () => {
```

Comment:
I don't think this is a good expectation - if anyone is interested in actions on a given microstate then it's likely they want to know **exact** actions associated with that particular transition/microstep and not in the list aggregating the actions since the beginning of that macrostep.

Reply:
This should be part of a follow-up PR, once we determine whether we want to keep actions in the state or not. Too big a change to make in this PR.